azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -44,3 +58,388 @@ def _run_op(self, op):
 
         else:
             super(UseSecurityClientStage, self)._run_op(op)
+
+
+class CommonProvisioningStage(PipelineStage):
+    @pipeline_thread.runs_on_pipeline_thread
+    def _get_error(self, provisioning_op, prov_op_name, error):
+        if error:
+            return error
+        elif 300 <= provisioning_op.status_code < 429:
+            logger.error(
+                "Received error with status code {status_code} for {prov_op_name} request operation".format(
+                    prov_op_name=prov_op_name, status_code=provisioning_op.status_code
+                )
+            )
+            logger.error("response body: {}".format(provisioning_op.response_body))
+            return exceptions.ServiceError(
+                "{prov_op_name} request returned a service error status code {status_code}".format(
+                    prov_op_name=prov_op_name, status_code=provisioning_op.status_code
+                )
+            )
+        else:
+            return None
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _decode_response(self, provisioning_op):
+        return json.loads(provisioning_op.response_body.decode("utf-8"))
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _get_registration_status(self, decoded_response):
+        return get_optional_element(decoded_response, "status")
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _get_operation_id(self, decoded_response):
+        return get_optional_element(decoded_response, "operationId")
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _form_complete_result(self, operation_id, decoded_response, status):
+        """
+        Create the registration result from the complete decoded json response for details regarding the registration process.
+        """
+        decoded_state = (get_optional_element(decoded_response, "registrationState"),)
+        registration_state = None
+        if decoded_state is not None:
```

Comment:
>decoded_state [](start = 11, length = 13)

The way you wrote this, decoded_state is guaranteed to never be None.  It might be (None,) but that is different.  Do you still have a reason to put this in a tuple and access it as decoded_state[0] or would it make more sense to remove the extra parens from your assignment above and access it as decoded_state?